lambda/debuginfo.ml

```diff
@@ -20,12 +20,30 @@ open Location
 module Scoped_location = struct
   type scope_item =
     | Sc_anonymous_function
-    | Sc_value_definition of string
-    | Sc_module_definition of string
-    | Sc_class_definition of string
-    | Sc_method_definition of string
+    | Sc_value_definition
+    | Sc_module_definition
+    | Sc_class_definition
+    | Sc_method_definition
 
-  type scopes = scope_item list
+  type scopes =
+    | Empty
+    | Cons of {item: scope_item; str: string; mutable str_fun: string}
+
+  module WeakString = Weak.Make(struct include String let hash = Hashtbl.hash end)
+
+  let memo_str = WeakString.merge (WeakString.create 128)
+
+  let str_fun = function
+    | Empty -> "(fun)"
+    | Cons r ->
+        if String.equal r.str_fun "" then r.str_fun <- memo_str (r.str ^ ".(fun)");
+        r.str_fun
```

Comment:
I'm not fond of this code:
- I don't like the use of `""` for an unitialized value, we should use a proper `string option` instead.
- Could we consider using `lazy` instead of explicit uninitialization then mutation?

(Using laziness increases the "runtime surface" again, but I think it adds less complexity than weak pointers. We could also consider dropping `str_fun` sharing completely, as the performance without it seem acceptable.)

Reply:
The whole point of this is to avoid some allocations.  If one doesn't really care, one could just always compute the `str_fun` in the `cons` function (at worse, that field will remain unused).

````ocaml
let cons item str =
  Cons {item; str; str_fun = lazy (str ^ ".(fun)"}
````

would feel a bit weird, and would not necessarily be more efficient than the naive approach (without the `lazy`).

The current code is not super nice, but the micro-optimization hack is very localized, and strictly better (performance-wise) than alternatives (naive approach, `string option`, `lazy`, or loosing sharing).

Reply:
I would prefer to keep the code simple, unless there is a clear evidence that the performance difference matters. Pre-computing `str_fun` would be fine, I think that the `lazy` approach is also reasonable, but I'm not in favor of having empty strings mean "uninitialized", it's not just worth the trouble when reading the code.

Reply:
Done (2d4d6d3166f16ff4880af6ad7c26fc021153c056).